Minimal solution for xgcd

albinahlback opened the pull request:
It ensures a minimal solution. However it does not ensure to get the same value as Julia's `gcdx` if both inputs are 0, for which we get (0, 0, 0) while Julia obtain (0, 1, 0).

It also feels like it could be optimized. Let me know if you have any suggestions.

Closes #948

Edit: This is about 40% slower than Nemo's `gcdx`. I definitely need to provide something faster.
Edit2: Nvm, this was about as fast (or like 1% faster) than the `BigInt`-implementation on big numbers. However, on small number around in [0, 2^10] it is about twice as fast.
Edit3: Conformed almost completely to Julia's `gcdx`.

fredrik-johansson commented:
Why are we not just calling mpz_gcdext directly?

I think our fmpz functions should agree with GMP unless there are strong reasons not to.

Does the new implementation agree with the "Canonical Bézout coefficients" here: https://fungrim.org/topic/Greatest_common_divisor/ ?

albinahlback commented:
How does this look? If it looks good, I think we should replace the current `xgcd.c` with `xgcd_minimal.c`.

fredrik-johansson commented:
Yes, I think we should do that. We should also add test code to check that it behaves as intended (we could compare with mpz_gcdext).

Just some comments about the code:

Aliasing doesn't work in the is_zero branches. This is easy to fix; you will just have to save the fmpz_sgn value in a register before you call fmpz_abs.

It is better to use _fmpz_promote / _fmpz_promote_val instead of mpz_init. In fact, you don't need the temporary mpz_t's at all if the output variables are not aliased with any inputs; just promote the outputs in-place.

To handle aliasing, I would maybe just check if any output is aliased with an input variable at the top of the function (after the is_zero checks), and in that case create temporary fmpz_t's and call the function recursively. I would not bother with something more clever because I think this function will nearly always be called with non-aliased arguments, so it's not a case we need to optimize for. We *do* want to avoid calling mpz_init / mpz_clear in the non-aliased case though because those functions are performance hogs.

You also don't need the temporary fmpz_t's for the single-limb absolute values; just use an mp_limb_t.

albinahlback commented:
Okay, so I tried to do that. Hopefully I wrote the function in the way you had in mind.

I also made some tests so the function works like your website wants it to (Haven't tested it yet though, so it may need some more work.)

fredrik-johansson commented:
Actually, I think the aliasing checks aren't needed if you just add a check for where both f and g are small simultaneously and then copy both to mp_limb_t's.

fredrik-johansson commented:
This looks fine to use for fmpz_xgcd now.

albinahlback commented:
What do you mean by saving *f and *g in registers?

The solution passes the tests now.

albinahlback commented:
I renamed the function and removed the original xgcd. Let me know if anything needs to be improved.

fredrik-johansson commented:
> What do you mean by saving *f and *g in registers?

Exactly what the code does now.

fredrik-johansson commented:
t-xgcd still references fmpz_xgcd_minimal.

albinahlback commented:
> t-xgcd still references fmpz_xgcd_minimal.

sorry, fixed it now

fredrik-johansson commented:
@wbhart will just need to enable CI.

wbhart commented:
The reason we weren't using GMP for small values is performance. The solutions we were returning should satisfy the inequalities in the GMP manual. If not, it was a bug.

I have no idea how that relates to the canonical bezout coefficients. Does GMP conform to those? It doesn't say so in the manual.

I am really quite worried about changing the definition of gcdx in Flint as it will change it in Nemo, and that definition is currently relied on. I don't have a proof that the minimal solution even satisfies the GMP inequalities, and just changing the Flint definition on a whim is not good behaviour on our part, since the existing definition has been there for years. I'd really like to examine this issue more carefully.

wbhart commented:
This PR doesn't even come with a change to the documentation to say what the new definitions would be.

Can we please:

1) ask the community what they want
2) document and test carefully
3) check performance

I'm ok with having an xgcd_minimal which does something different to the current xgcd. But the current xgcd was implemented a certain way for a reason.

albinahlback commented:
So I think I fixed the speed for small pairs. Currently it is only 3% slower than `fmpz_xgcd` on the tests I tried.

I also added some more tests.

I also fixed a better documentation.

I also want to mention that I added `FLINT_SGN` since I found it very useful.

This should be ready to merge, unless you can find improvements!

wbhart commented on fmpz/test/t-xgcd_minimal.c:
```diff
@@ -0,0 +1,627 @@
+    }
+
+
+    _fmpz_promote_val(a);
```
There's no need to test promoted values work. There's no expectation that any fmpz function accept promoted fmpz's. Promoted values are for internal use.

wbhart commented:
Could you also add tests for xgcd_minimal(a, ab) and xgcd_minimal(ab, a)

Also, is there a test that actually tests the minimal condition somewhere? |s| + |t| is minimal? It should be possible to test by checking that increasing s +/- b and t -/+ a don't give smaller |s| + |t|.

wbhart commented:
Oh wait, now you are using canonical bezout coefficients instead of your original definition?

The current description is not very complete. It say that "often" some inequalities hold. It's not very useful unless you state precisely the definition, or give a reference. Also, wouldn't xgcd_canonical_bezout be a better name in this case?

Can you also add a random test which generates random inputs and then checks all those inequalities to make sure they hold.

albinahlback commented:
Sorry for all the back-and-forth. I have commented the test and structured it according to the documentation I rewrote.

albinahlback commented:
@wbhart Should `FLINT_SGN` be documented?

tthsqe12 commented:
I am starting to think that these kind of macros are a bad idea.
@fredrik-johansson what is going on here
https://github.com/fredrik-johansson/arb/blob/master/arf.h#L443

tthsqe12 commented:
Can this new fmpz_xgcd function be summarized as
make it agree with mpz_gcdext and as fast as the old fmpz_xgcd in the small case?

albinahlback commented:
@tthsqe12 I believe so.

tthsqe12 commented:
If you are into micro optimizations, both of the gcd's in `ulong_extras/gcd.c` are much slower than just calling `mpn_gcd_11`.
Also, `ulong_extras/n_gcdinv` is not particularly clever either, and probably can be sped up with a binary approach.

fredrik-johansson commented:
> I am starting to think that these kind of macros are a bad idea.
> @fredrik-johansson what is going on here
> https://github.com/fredrik-johansson/arb/blob/master/arf.h#L443

This is for returning the absolute value of an slong as a ulong, correctly handling the WORD_MIN case. I don't remember where or how, but there was clearly some case where FLINT_ABS didn't do the job.

fredrik-johansson commented:
> If you are into micro optimizations, both of the gcd's in `ulong_extras/gcd.c` are much slower than just calling `mpn_gcd_11`.
> Also, `ulong_extras/n_gcdinv` is not particularly clever either, and probably can be sped up with a binary approach.

Interesting. These functions are pretty crucial for rational and modular arithmetic so we should make sure they are as fast as possible.

fredrik-johansson commented:
By the way, the "canonical Bezout coefficients" are not canonical in any deep sense; this is just the convention GMP happens to use.

wbhart commented:
@tthsqe12 Where is mpn_gcd_11 documented? I see only mpn_gcd_1 in their manual.

Note that we cannot just use their code, due to license. And if it isn't in the documented interface, we can't use it either.

tthsqe12 commented:
If mpn_gcd_11 isn't documented, mpn_gcd_1 is just fine.

wbhart commented:
By fine, do you mean "faster than what we already have"?

tthsqe12 commented:
probably :)

wbhart commented:
You really think their version will be faster than calling assembly instructions for counting leading zeros? I find this hard to believe.

albinahlback commented:
I did improve the method. In the new commit I provided the method `xgcd_canonical_bezout`, which is faster than the first method. Moreover it seems like it is way faster than `xgcd` on bigger instances. Using https://gist.github.com/albinahlback/b4a5ffc01c2d4b1b9629cf30187a4e88 this is some of the results I obtained:

```
$ ./a.out 23058430092136939 23058430092146959
# first argument is minimum in for loop
# second is maximum
# time is in milliseconds, I believe
cpu: 10972 # new version
wall: 10979

cpu: 11185 # my old version
wall: 11197

cpu: 16253 # original xgcd
wall: 16260

$ ./a.out 0 15000
cpu: 20474
wall: 20510

cpu: 20826
wall: 20833

cpu: 20259
wall: 20264

$ ./a.out 10000000000 10000003000
cpu: 869
wall: 872

cpu: 879
wall: 882

cpu: 1300
wall: 1304
```

albinahlback commented:
Let me know which one to choose and if there is further improvements that one, so we can merge this! :-)

albinahlback commented:
Okay, I changed `mpn_gcdext_1` since it was the fastest and was able to pass all the tests.

I think this is as good as it gets.

tthsqe12 commented:
If there are some timings that you would like to preserve, put some code in the profile directory. That way, you can be sure they are tested against future "improvements".

wbhart commented:
It would actually be useful to do some timings on some (carefully chosen) randomised examples.

albinahlback commented:
I haven't set `FLINT_CLOCKSPEED`, but here is one result which describes the result on my Haswell 4770k:
```
fmpz_xgcd (small size):
  min time is 16.073 cycles
  max time is 18.479 cycles

fmpz_xgcd (mixed size):
  min time is 24.486 cycles
  max time is 24.934 cycles

fmpz_xgcd (big size):
  min time is 138.513 cycles
  max time is 140.990 cycles

fmpz_xgcd_canonical_bezout (small size):
  min time is 14.623 cycles
  max time is 15.561 cycles

fmpz_xgcd_canonical_bezout (mixed size):
  min time is 19.420 cycles
  max time is 19.535 cycles

fmpz_xgcd_canonical_bezout (big size):
  min time is 132.021 cycles
  max time is 135.418 cycles

```

wbhart commented:
From my perspective, this PR looks good now. If no one has any objections to it, I'd like to merge it.

@fredrik-johansson @tthsqe12

tthsqe12 commented:
none here
